### PR TITLE
Add support for Class symbols

### LINKED ISSUE #26071: tsymbol of BObjectType is wrong for classes from external modules
**Description:**
tsymbol in BObjectType should be BClassSymbol if the variable is a class. This works for classes defined in the same module, but when the class is defined in another module the tsymbol is BObjectTypeSymbol.

**Steps to reproduce:**

1. Have the following code in a bal file.

```
import ballerina/cache;

# Test function
#
# + prsn - The stream
# + cacheCls - The cache class from cache module
public isolated function testPerson(Person prsn, cache:Cache cacheCls) returns string?|Person {
}

# Represents a person object
#
# + name - Name of the person
# + age - Age of the person in years
# + address - Address of the person
public class Person {
    public string name = "";
    public int age;
    public string address;

    # Get address of the person
    #
    # + return - New address of the person
    public function getAddress() returns string {
        return "new" ;
    }
}
```
2. Get the BLangPackage
3. Check functions->requiredParams->cacheCls->type->tsymbol which is BObjectTypeSymbol
4. Check functions->requiredParams->prsn->type->tsymbol which is BClassSymbol

**Affected Versions:**

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### LINKED ISSUE #26251: Symbol at cursor functionality for type inits is incomplete
It seems the functionality of the symbol at cursor API for type inits is not entirely correct. If the cursor is at an explicitly stated type reference, it should return the symbol for the init method. The current behaviour doesn't consider the init method at all. Even if the init method is considered, there's the case where there may not be a user-defined init method.

### LINKED ISSUE #26265: Need to differentiate between classes and abstract objects
**Description:**
Using the command `typeDescriptor().get().typeDescriptor().get().kind()` returns as `OBJECTS` for both classes and abstarct objects. We need a way to differentiate between classes and abstract objects for API Doc generation.

**Steps to reproduce:**

**Affected Versions:**

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**
@pubudu91

### PR DESCRIPTION
**Pull request by pubudu91**
## Purpose
This PR adds a separate symbol for classes.

Fix #26265
Fix #26071
Fix #26251

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Review by nadeeshaan — CHANGES_REQUESTED** [review]

File: tests/ballerina-compiler-api-test/src/test/resources/test-src/class_symbols_test.bal [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    public function getFullName() returns string {
+        return self.fname + " " + self.lname;
+    }
+}
+
+function test() {
+    Person1 p1 = new("Pubudu", "Fernando");
+    Person2 p2 = new;
+    p2 = new Person2();
+}
```
Comment (SupunS) [inline tests/ballerina-compiler-api-test/src/test/resources/test-src/class_symbols_test.bal:44]: Missing eof

File: compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/BallerinaSemanticModel.java [hunk 2]
```diff
@@ -111,7 +115,14 @@ public BallerinaSemanticModel(BLangPackage bLangPackage, CompilerContext context
             return Optional.empty();
         }
 
+        if (isTypeSymbol(symbolAtCursor) && !PositionUtil.withinBlock(position, symbolAtCursor.pos)) {
+            ModuleID moduleID = new BallerinaModuleID(symbolAtCursor.pkgID);
+            return Optional.of(new BallerinaTypeReferenceTypeSymbol(moduleID, symbolAtCursor.type,
+                                                                    symbolAtCursor.getName().getValue()));
+        }
+
```
[the remark this answers is not in the document]
Comment (pubudu91) [reply compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/BallerinaSemanticModel.java:123]: Can't really do that right? What's being checked here is the following case:
```ballerina
type Person record {|
    string name;
|};

function bar() {
    <cursor>Person p = { name: "Pubudu" };
}
```
In this case, I want to return a `TypeReferenceTypeSymbol` instance, not a `TypeDefinitionSymbol`. To check which one to return, have to check the `BSymbol`'s position against the user provided position.

[the remark this answers is not in the document]
Comment (SupunS) [reply compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/BallerinaSemanticModel.java:123]: In the AST we would know its a type-reference right? 
for e.g.: in `public void visit(BLangSimpleVariable varNode) {}` method, we know what we need to return is a `TypeReferenceTypeSymbol`

[the remark this answers is not in the document]
Comment (pubudu91) [reply compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/BallerinaSemanticModel.java:123]: Yeah, but the thing is, we return a `BSymbol` from the `SymbolFinder`, not the node.

**Review by SupunS — APPROVED** [review]

**Review by nadeeshaan — APPROVED** [review]

**Pull request merged**